### PR TITLE
 #171380767 edit open travel request

### PR DESCRIPTION
**Pull request by Ugizwenayo-Divine**
#### What does this PR do?
Creates API endpoint used to edit travel requests
#### Description of Task to be completed?
- create API endpoint at ***PATCH/api/trips/:requestId***
- add updateTripRequest function in src/controllers/request.controller.js
- add updateTripRequest function in src/services/request.service.js
- create requesterChecker.js file in src/middlewares
- create tests to test edit travel request functionality
#### How should this be manually tested?
- After cloning the project repo, run `npm i` inside the cloned repo folder
- Make sure to rename .env.example to .env, 
- In .env file, provide all values for variables under `#Development database`  and `#General`
- Finally, run the following commands respectively:
`git fetch`
`git checkout  ft-update-open-request-171380767`
`npm i`
`npx sequelize-cli db:create`
`npx sequelize-cli db:migrate`
`npm run start`
- To edit travel request, send the following data in your request body to API at `PATCH http://<your_server_hostname>:<server_port>/api/trips/:requestId` endpoint:
```json
   "travelFrom": "kigali",
    "travelTo": "kibuye",
    "travelDate": "2020-04-30",
    "travelReason": "business",
    "travelType": "one-way",
    "accommodation": true
```
#### Any background context you want to provide?
You need to create an account and use the obtained token as an `authorization header` in your requests to the API.
#### What are the relevant pivotal tracker stories?
[#171380767](https://www.pivotaltracker.com/story/show/171380767)
#### Screenshots (if appropriate)

Capturedi1

Capturedi2

Capturedi3

Capturedi4

Capturedi5

### CODE REVIEW

File: src/middlewares/requestChecker.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+   * get request id from request parameter, checks if the request exist or not
+   */
+  const isRequestOpenIsRequestYours = async (req, res, next) => {
+    const tokenDecoded = req.sessionUser;
+    const owner = req.requestOwner;
+    const status = req.requestStatus;
+    if (tokenDecoded.id !== owner) {
+        return errorResponse(res, badRequest, notYourRequest);
+    }
+    if (status !== 'pending') {
```
Comment (since9629) [inline src/middlewares/requestChecker.js]: could you use globals declared in `src/utils/tripRequestsStatus.util.js`

File: src/utils/customMessages.js [hunk 2]
```diff
@@ -97,6 +97,11 @@ const messages = {
   viewStatsUnauthorized: 'Permission denied trying to view trips stats of another user, you must be the owner to view the stats',
   viewStatsNoRequesterId: 'Please specify a valid value for the requester id, it must be a positive number',
   requesterNotRegistered: 'Sorry, we could not find a user account associated with that requesterId',
+  requestUpdated: 'The request is updated successfully',
+  notOpenRequest: 'You can not updated rejected or approved travel request',
```
Comment (the22mastermind) [inline src/utils/customMessages.js]: Could you correct this to `You cannot update a rejected or approved travel request`?

**Review by the22mastermind — COMMENTED** [review]
Good job @Ugizwenayo-Divine 
Could you please make sure the route specified in the PR description matches the one in your code?
Check out also the comment I left on customMessages.js

**Review by since9629 — APPROVED** [review]

**Review by descholar-ceo — APPROVED** [review]
Nice job @Ugizwenayo-Divine

**Review by the22mastermind — APPROVED** [review]

File: tests/other/request.test.js [hunk 3]
```diff
@@ -1367,6 +1381,148 @@ describe('Trips stats', () => {
...
+      .request(server)
+      .patch(`/api/trips/${requestId}`)
+      .set('Authorization', authToken)
+      .send(travelUpdated)
+      .end((err, res) => {
+        if (err) done(err);
+        expect(res).to.have.status(badRequest);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('error').to.equal(duplicateTripRequest);
+        done();
```
Comment (kipropbrian) [inline tests/other/request.test.js]: The description of this test is not correct.

File: tests/other/request.test.js [hunk 4]
```diff
@@ -1367,6 +1381,148 @@ describe('Trips stats', () => {
...
+        expect(res).to.have.status(ok);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('message').to.equal(requestUpdated);
+        done();
+      });
+  });
+  it('should not update approved or rejected travel request', (done) => {
+    chai
+      .request(server)
+      .patch('/api/trips/13')
```
Comment (kipropbrian) [inline tests/other/request.test.js]: Hard coded id's may introduce errors in the future. Same as the one in @descholar-ceo  code.

File: tests/other/request.test.js [hunk 5]
```diff
@@ -1367,6 +1381,148 @@ describe('Trips stats', () => {
...
+        expect(res).to.have.status(badRequest);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('error').to.equal(emptyUpdate);
+        done();
+      });
+  });
+  it('should not update open travel request which does not exist', (done) => {
+    chai
+      .request(server)
+      .patch('/api/trips/12')
```
Comment (kipropbrian) [inline tests/other/request.test.js]: same here. You can use 50000 or a large number.

**Review by kipropbrian — COMMENTED** [review]
Work on the changes on the tests and then it's good to go.

**Pull request merged**